File: c++/array/q414.cpp
//NOTE: Trakcing the largest 3 numbers
int thirdMax1(vector<int>& nums) {
    set<int> top3;
    for(int i = 0; i < nums.size(); i++){
        top3.insert(nums[i]);
        if(top3.size() > 3)
            top3.erase(top3.begin());
    }
    return top3.size() == 3 ? *top3.begin() : *top3.rbegin();
}


int thirdMax2(vector<int>& nums) {
    long f = LONG_MIN;
    long s = LONG_MIN;
    long t = LONG_MIN;

    for(int i = 0; i < nums.size(); i++){
        int temp = nums[i];
        if(f == temp || s == temp || t == temp)
            continue;
        if(temp > f){
            t = s;
            s = f;
            f = temp;
        }
        else if(temp > s){
            t = s;
            s = temp;
        }
        else if(temp > t){
            t = temp;
        }
    }

    return t == LONG_MIN ? f : t;
}

File: python/array/q414.py
"""
Given a non-empty array of integers, return the third maximum number in this array. If it does not exist, return the maximum number. The time complexity must be in O(n).

Example 1:
Input: [3, 2, 1]

Output: 1

Explanation: The third maximum is 1.
Example 2:
Input: [1, 2]

Output: 2

Explanation: The third maximum does not exist, so the maximum (2) is returned instead.
Example 3:
Input: [2, 2, 3, 1]

Output: 1

Explanation: Note that the third maximum here means the third maximum distinct number.
Both numbers with value 2 are both considered as second maximum.
"""
def thirdMax(self, nums):
    """
    :type nums: List[int]
    :rtype: int
    """
    top3 = [float("-inf"),float("-inf"),float("-inf")]
    for val in nums:
        if val > top3[0]:
            top3 = [val,top3[0],top3[1]]
        elif val < top3[0] and val > top3[1]:
            top3 = [top3[0],val,top3[1]]
        elif val < top3[1] and val > top3[2]:
            top3 = [top3[0],top3[1],val]

    if top3[2] == float("-inf"):
        return top3[0]
    else:
        return top3[2]
